web_sdk/sdk_rewriter.dart

```diff
@@ -17,78 +17,159 @@ final ArgParser argParser = ArgParser()
+  AllReplacer("import 'package:meta/meta.dart';", ''),
+  AllReplacer('@required', ''),
+  AllReplacer('@protected', ''),
+  AllReplacer('@mustCallSuper', ''),
+  AllReplacer('@immutable', ''),
+  AllReplacer('@visibleForTesting', ''),
 ];
 
 // Rewrites the "package"-style web ui library into a dart:ui implementation.
 // So far this only requires a replace of the library declarations.
 void main(List<String> arguments) {
   final ArgResults results = argParser.parse(arguments);
-  final Directory directory = Directory(results['output-dir']);
-  final String inputDirectoryPath = results['input-dir'];
-  for (String inputFilePath in results['input']) {
+  final Directory directory = Directory(results['output-dir'] as String);
+  final String inputDirectoryPath = results['input-dir'] as String;
+  for (final String inputFilePath in results['input'] as Iterable<String>) {
     final File inputFile = File(inputFilePath);
     final File outputFile = File(path.join(
         directory.path, inputFile.path.substring(inputDirectoryPath.length)))
       ..createSync(recursive: true);
     String source = inputFile.readAsStringSync();
-    final List<List<String>> replacementPatterns = <List<String>>[];
+    final List<Replacer> replacementPatterns = <Replacer>[];
     replacementPatterns.addAll(sharedPatterns);
-    if (results['ui']) {
+    if (results['ui'] as bool) {
       replacementPatterns.addAll(uiPatterns);
-    } else if (results['engine']) {
-      replacementPatterns.addAll(enginePatterns);
+    } else if (results['engine'] as bool) {
+      if (inputFilePath.endsWith('lib/src/engine.dart')) {
+        replacementPatterns.addAll(engineLibraryPatterns);
+      } else {
+        source = _preprocessEnginePartFile(source);
+        replacementPatterns.addAll(enginePartsPatterns);
+      }
     }
-    for (List<String> patterns in replacementPatterns) {
-      source = source.replaceAll(patterns.first, patterns.last);
+    for (final Replacer replacer in replacementPatterns) {
+      source = replacer.perform(source);
     }
     outputFile.writeAsStringSync(source);
     if (results['stamp'] != null) {
-      File(results['stamp']).writeAsStringSync("stamp");
+      File(results['stamp'] as String).writeAsStringSync('stamp');
     }
   }
 }
+
+final RegExp _dartLangVersion = RegExp(r'\n// @dart = (.*)');
+
+String _preprocessEnginePartFile(String source) {
+  if (source.contains('\npart of engine;')) {
+    // The file hasn't been migrated yet.
+    // Do nothing.
+  } else if (_dartLangVersion.hasMatch(source)) {
+    // The file contains a dart language version. We should insert the part
+    // directive after it.
+    source = source.replaceFirstMapped(_dartLangVersion, (Match match) {
+      return '${match.group(0)}\npart of engine;';
+    });
+  } else {
+    // The file doesn't have a dart language version, so we just insert the part
+    // directive at the beginning.
+    source = 'part of engine;\n' + source;
```

Comment:
It's strange that some files don't have the language version. Can we make them all consistent?

Reply:
All files have the language version. They can't not have it. The build would fail. But I was trying to write the code assuming we may remove language versions at some point.